parquet/src/main/java/org/apache/iceberg/data/parquet/InternalReader.java

```diff
@@ -0,0 +1,207 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data.parquet;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.parquet.ParquetValueReader;
+import org.apache.iceberg.parquet.ParquetValueReaders;
+import org.apache.iceberg.parquet.ParquetValueReaders.StructReader;
+import org.apache.iceberg.types.Types.StructType;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+public class InternalReader extends BaseParquetReaders<StructLike> {
+
+  private static final InternalReader INSTANCE = new InternalReader();
+
+  private InternalReader() {}
+
+  public static ParquetValueReader<StructLike> buildReader(
+      Schema expectedSchema, MessageType fileSchema) {
+    return INSTANCE.createReader(expectedSchema, fileSchema);
+  }
+
+  public static ParquetValueReader<StructLike> buildReader(
+      Schema expectedSchema, MessageType fileSchema, Map<Integer, ?> idToConstant) {
+    return INSTANCE.createReader(expectedSchema, fileSchema, idToConstant);
+  }
+
+  @Override
+  protected ParquetValueReader<StructLike> createStructReader(
+      List<Type> types, List<ParquetValueReader<?>> fieldReaders, StructType structType) {
+    return new ParquetStructReader(types, fieldReaders, structType);
+  }
+
+  @Override
+  protected LogicalTypeAnnotation.LogicalTypeAnnotationVisitor<ParquetValueReader<?>>
+      logicalTypeReaderVisitor(
+          ColumnDescriptor desc,
+          org.apache.iceberg.types.Type.PrimitiveType expected,
+          PrimitiveType primitive) {
+    return new LogicalTypeAnnotationParquetValueReaderVisitor(desc, expected, primitive);
+  }
+
+  @Override
+  protected ParquetValueReaders.PrimitiveReader<?> fixedReader(ColumnDescriptor desc) {
+    return new ParquetValueReaders.BytesReader(desc);
+  }
+
+  @Override
+  protected ParquetValueReaders.PrimitiveReader<?> int96Reader(ColumnDescriptor desc) {
+    // normal handling as int96
+    return new ParquetValueReaders.UnboxedReader<>(desc);
```

Comment:
This isn't correct. The unboxed reader will return a `Binary` for int96 columns. Instead, this needs to use the same logic as the Spark reader (which also uses the internal representation):

```java
  private static class TimestampInt96Reader extends UnboxedReader<Long> {
    TimestampInt96Reader(ColumnDescriptor desc) {
      super(desc);
    }

    @Override
    public Long read(Long ignored) {
      return readLong();
    }
    @Override
    public long readLong() {
      final ByteBuffer byteBuffer =
          column.nextBinary().toByteBuffer().order(ByteOrder.LITTLE_ENDIAN);
      return ParquetUtil.extractTimestampInt96(byteBuffer);
    }
  }
```

You can move that class into the `parquet` package to share it.